src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs

```diff
@@ -175,11 +194,11 @@ internal AssemblyQualifiedTypeName DecodeTypeName(bool isTypeArgument = false, b
                                 if (isGenericTypeName && typeArguments == null)
                                 {
                                     Advance();
-                                    if (arrayRanksBuilder != null)
+                                    if (arrayRanksBuilder != null || pointerCount > 0)
                                     {
                                         // Error case, array shape must be specified at the end of the type name.
                                         // Process as a regular character and continue.
-                                        typeNamebuilder.Append('[');
+                                        typeNamebuilder.Append(c);
```

Comment:
I was concerned about copy-paste errors.  And, conceptually, appending the just-consumed character seemed clearer than appending a literal.